plugins/providers/gcs/config.go

```diff
@@ -0,0 +1,195 @@
+		validator:      validator.New(),
+		crypto:         crypto,
+	}
+}
+
+func (c *Config) ParseAndValidate() error {
+	return c.parseAndValidate()
+}
+
+func (c *Config) parseAndValidate() error {
+	if c.valid {
+		return nil
+	}
+
+	validationError := []error{}
+
+	credentials, err := c.validateCredentials(c.ProviderConfig.Credentials)
+	if err != nil {
+		validationError = append(validationError, err)
+	} else {
+		c.ProviderConfig.Credentials = credentials
+	}
+	ctx := context.TODO()
+	saKey := credentials.ServiceAccountKey
+	fmt.Printf("saKey: %v\n", saKey)
+	client, err := storage.NewClient(ctx, option.WithCredentialsJSON([]byte(saKey)))
+	if err != nil {
+		return fmt.Errorf("initialising gcs client: %w", err)
+	}
+	defer client.Close()
+
+	for _, r := range c.ProviderConfig.Resources {
+		if err := c.validateResourceConfig(r); err != nil {
+			validationError = append(validationError, err)
+		}
+	}
+
+	if len(validationError) > 0 {
+		errorStrings := []string{}
+		for _, err := range validationError {
+			errorStrings = append(errorStrings, err.Error())
+		}
+		return errors.New(strings.Join(errorStrings, "\n"))
+	}
+
+	c.valid = true
+	return nil
+}
+
+func (c *Config) validateCredentials(value interface{}) (*Credentials, error) {
+	var credentials Credentials
+	if err := mapstructure.Decode(value, &credentials); err != nil {
+		return nil, err
+	}
+
+	if err := c.validator.Struct(credentials); err != nil {
+		return nil, err
+	}
+
+	saKeyJson, err := base64.StdEncoding.DecodeString(credentials.ServiceAccountKey)
+	if err != nil {
+		return nil, err
+	}
+
+	credentials.ServiceAccountKey = string(saKeyJson)
+
+	return &credentials, nil
+}
+
+func (c *Config) validateResourceConfig(resource *domain.ResourceConfig) error {
+	resourceTypeValidation := fmt.Sprintf("oneof=%s", ResourceTypeBucket)
+	if err := c.validator.Var(resource.Type, resourceTypeValidation); err != nil {
+		return fmt.Errorf("validating resource type: %w", err)
+	}
+
+	for _, role := range resource.Roles {
+		for i, permission := range role.Permissions {
+			if permissionConfig, err := c.validatePermission(resource.Type, permission); err != nil {
+				return fmt.Errorf("validating permissions: %w", err)
+			} else {
+				role.Permissions[i] = permissionConfig
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Config) validatePermission(resourceType string, value interface{}) (*Permission, error) {
+	permissionConfig, ok := value.(string)
+	if !ok {
+		return nil, ErrInvalidPermissionConfig
+	}
+
+	var pc Permission
+	if err := mapstructure.Decode(permissionConfig, &pc); err != nil {
+		return nil, err
+	}
+
+	var nameValidation string
+	if resourceType == ResourceTypeBucket {
+		nameValidation = "oneof=roles/storage.admin roles/storage.legacyBucketOwner roles/storage.legacyBucketReader roles/storage.legacyBucketWriter roles/storage.legacyObjectOwner roles/storage.legacyObjectReader roles/storage.objectAdmin roles/storage.objectCreator roles/storage.objectViewer"
```

Comment:
can we define these validation strings as const?

Reply:
yep, will do it